File: ecomate-mobile-backend-service/ecomate-mobile-backend-service-master/service/news.go
package service

import (
	"context"

	db "github.com/Reigenleif/ecomate-mobile-backend-service/internal/db"
	"github.com/Reigenleif/ecomate-mobile-backend-service/internal/token_service"
	models "github.com/Reigenleif/ecomate-mobile-backend-service/models"
	proto "github.com/Reigenleif/ecomate-mobile-backend-service/proto"
	"github.com/gofrs/uuid"
)

type NewsService struct {
	proto.UnimplementedNewsServiceServer
}

func (s *NewsService) GetNewsList(ctx context.Context, req *proto.GetNewsListRequest) (*proto.NewsListResponse, error) {
	rows, err := db.GetDB().Query(ctx, "SELECT * FROM public.\"News\"")
	if err != nil {
		return nil, err
	}

	var newsDbList []models.News
	for rows.Next() {
		var newsDb models.News
		err := rows.Scan(&newsDb.ID, &newsDb.CreatedAt, &newsDb.UpdatedAt, &newsDb.Title, &newsDb.ImageUrl, &newsDb.Content, &newsDb.IsPublished)
		if err != nil {
			return nil, err
		}
		newsDbList = append(newsDbList, newsDb)
	}

	var newsList []*proto.News
	for _, item := range newsDbList {
		newsList = append(newsList, &proto.News{
			Id:       item.ID.String,
			Title:    item.Title.String,
			ImageUrl: item.ImageUrl.String,
			Content:  item.Content.String,
		})
	}

	return &proto.NewsListResponse{
		NewsList: newsList,
	}, nil
}

func (s *NewsService) GetNewsCommentList(ctx context.Context, req *proto.GetNewsCommentListRequest) (*proto.NewsCommentListResponse, error) {
	rows, err := db.GetDB().Query(ctx, "SELECT * FROM public.\"NewsComment\" WHERE \"newsId\" = '"+req.NewsId+"';")
	if err != nil {
		return nil, err
	}

	newsCommentDbList := []models.NewsComment{}
	for rows.Next() {
		var newsCommentDb models.NewsComment
		err := rows.Scan(&newsCommentDb.ID, &newsCommentDb.CreatedAt, nil, &newsCommentDb.Content, &newsCommentDb.NewsID, &newsCommentDb.UserID)
		if err != nil {
			return nil, err
		}
		newsCommentDbList = append(newsCommentDbList, newsCommentDb)
	}

	var newsCommentList []*proto.NewsComment
	for _, item := range newsCommentDbList {
		newsCommentList = append(newsCommentList, &proto.NewsComment{
			Id:      item.ID.String,
			CreatedAt: item.CreatedAt.String(),
			Content: item.Content.String,
			NewsId:  item.NewsID.String,
			UserId:  item.UserID.String,
		})
	}

	return &proto.NewsCommentListResponse{
		NewsCommentList: newsCommentList,
	}, nil
}

func (s *NewsService) CreateNewsComment(ctx context.Context, req *proto.CreateNewsCommentRequest) (*proto.GeneralStatusResponse, error) {
	userClaims, err := token_service.CheckToken(ctx)
	if err != nil {
		return nil, err
	}

	randId := uuid.Must(uuid.NewV4()).String()
	_, errExec := db.GetDB().Exec(ctx, "INSERT INTO public.\"NewsComment\" (id, content, \"newsId\", \"userId\") VALUES ($1, $2, $3, $4)", randId, req.Content, req.NewsId, userClaims.Id)
	if errExec != nil {
		return nil, errExec
	}

	return &proto.GeneralStatusResponse{
		Success: true,
	}, nil
}

func (s *NewsService) UpdateNewsComment(ctx context.Context, req *proto.UpdateNewsCommentRequest) (*proto.GeneralStatusResponse, error) {
	userClaims, err := token_service.CheckToken(ctx)
	if err != nil {
		return nil, err
	}

	_, errExec := db.GetDB().Exec(ctx, "UPDATE public.\"NewsComment\" SET content = $1 WHERE id = $2 AND \"userId\" = $3", req.Content, req.CommentId, userClaims.Id)
	if err != nil {
		return nil, errExec
	}

	return &proto.GeneralStatusResponse{
		Success: true,
	}, nil
}

func (s *NewsService) DeleteNewsComment(ctx context.Context, req *proto.DeleteNewsCommentRequest) (*proto.GeneralStatusResponse, error) {
	userClaims, err := token_service.CheckToken(ctx)
	if err != nil {
		return nil, err
	}

	_, errExec := db.GetDB().Exec(ctx, "DELETE FROM public.\"NewsComment\" WHERE id = $1 AND \"userId\" = $2", req.CommentId, userClaims.Id)
	if err != nil {
		return nil, errExec
	}

	return &proto.GeneralStatusResponse{
		Success: true,
	}, nil
}

File: ecomate-mobile-ui/lib/provider/news.dart
import 'package:ecomate/proto/main.pb.dart';
import 'package:ecomate/services/news.dart';
import 'package:flutter/material.dart';

class NewsComment {
  final String id;
  final String content;
  final String newsId;

  final String userId;

  NewsComment(
      {required this.id,
      required this.content,
      required this.newsId,
      required this.userId});
}

class NewsItem {
  final String id;
  final String title;
  final String content;
  final String imageUrl;
  late List<NewsComment> newsCommentList;

  setNewsCommentList(List<NewsComment> newsCommentList) {
    this.newsCommentList = newsCommentList;
  }

  NewsItem(
      {required this.id,
      required this.title,
      required this.content,
      required this.imageUrl});
}

class NewsState extends ChangeNotifier {
  List<NewsItem> newsItemList = [];

  // Pure Functions for State
  void setNewsItemList(List<NewsItem> newsItemList) {
    this.newsItemList = newsItemList;
    notifyListeners();
  }

  void setNewsCommentList(String newsId, List<NewsComment> newsCommentList) {
    for (var newsItem in newsItemList) {
      if (newsItem.id == newsId) {
        newsItem.setNewsCommentList(newsCommentList);
      }
    }
    notifyListeners();
  }

  // Async GRPC Calls
  Future<void> queryNewsList() async {
    final request = GetNewsListRequest();
    final response =
        await NewsService.instance.newsServiceClient.getNewsList(request);
    List<NewsItem> newsItemList = [];
    for (var newsItem in response.newsList) {
      newsItemList.add(NewsItem(
          id: newsItem.id,
          title: newsItem.title,
          content: newsItem.content,
          imageUrl: newsItem.imageUrl));
    }
    setNewsItemList(newsItemList);
  }

  Future<void> queryNewsCommentList(String newsId) async {
    final request = GetNewsCommentListRequest(newsId: newsId);
    final response = await NewsService.instance.newsServiceClient
        .getNewsCommentList(request);
    List<NewsComment> newsCommentList = [];
    for (var newsComment in response.newsCommentList) {
      newsCommentList.add(NewsComment(
          id: newsComment.id,
          content: newsComment.content,
          newsId: newsComment.newsId,
          userId: newsComment.userId));
    }
    setNewsCommentList(newsId, newsCommentList);
  }

  Future<void> mutateCreateNewsComment(String newsId, String content) async {
    final request = CreateNewsCommentRequest(newsId: newsId, content: content);
    await NewsService.instance.newsServiceClient.createNewsComment(request);
    await queryNewsCommentList(newsId);
  }

  Future<void> mutateUpdateNewsComment(
      String newsId, String newsCommentId, String content) async {
    final request = UpdateNewsCommentRequest(
        newsId: newsId, commentId: newsCommentId, content: content);
    await NewsService.instance.newsServiceClient.updateNewsComment(request);
    await queryNewsCommentList(newsId);
  }

  Future<void> mutateDeleteNewsComment(
      String newsId, String newsCommentId) async {
    final request =
        DeleteNewsCommentRequest(newsId: newsId, commentId: newsCommentId);
    await NewsService.instance.newsServiceClient.deleteNewsComment(request);
    await queryNewsCommentList(newsId);
  }

  // Pure Functions for UI
  List<NewsItem> getNewsItemList() {
    return newsItemList;
  }

  List<NewsComment> getNewsCommentList(String newsId) {
    for (var newsItem in newsItemList) {
      if (newsItem.id == newsId) {
        return newsItem.newsCommentList;
      }
    }
    return [];
  }
}
